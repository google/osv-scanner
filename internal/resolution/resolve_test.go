package resolution_test

import (
	"cmp"
	"context"
	"slices"
	"testing"

	"deps.dev/util/resolve"
	"deps.dev/util/resolve/dep"
	"github.com/google/osv-scanner/internal/resolution"
	"github.com/google/osv-scanner/internal/resolution/clienttest"
	"github.com/google/osv-scanner/internal/resolution/manifest"
	"github.com/google/osv-scanner/internal/testutility"
)

func checkResult(t *testing.T, result *resolution.Result) {
	t.Helper()
	snap := testutility.NewSnapshot()
	snap.MatchText(t, result.Graph.String())

	type minimalVuln struct {
		ID               string
		DevOnly          bool
		ProblemChains    [][]resolve.Edge
		NonProblemChains [][]resolve.Edge
	}

	minVulns := make([]minimalVuln, len(result.Vulns))
	for i, v := range result.Vulns {
		minVulns[i] = minimalVuln{
			ID:               v.OSV.ID,
			DevOnly:          v.DevOnly,
			ProblemChains:    make([][]resolve.Edge, len(v.ProblemChains)),
			NonProblemChains: make([][]resolve.Edge, len(v.NonProblemChains)),
		}
		for j, c := range v.ProblemChains {
			minVulns[i].ProblemChains[j] = c.Edges
		}
		for j, c := range v.NonProblemChains {
			minVulns[i].NonProblemChains[j] = c.Edges
		}
	}
	slices.SortFunc(minVulns, func(a, b minimalVuln) int {
		return cmp.Compare(a.ID, b.ID)
	})
	snap.MatchJSON(t, minVulns)
}

func TestResolve(t *testing.T) {
	t.Parallel()

	aliasType := func(knownAs string) dep.Type {
		t.Helper()
		typ := dep.NewType()
		typ.AddAttr(dep.KnownAs, knownAs)

		return typ
	}

	type requirement struct {
		name    string
		version string
		typ     dep.Type
		groups  []string
	}
	tests := []struct {
		name         string
		version      string
		system       resolve.System
		universe     string
		requirements []requirement
		opts         resolution.ResolveOpts
	}{
		{
			name:     "simple", // simple root -> dependency -> vuln
			version:  "1.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/basic-universe.yaml",
			requirements: []requirement{
				{
					name:    "dependency",
					version: "^1.0.0",
					groups:  []string{"dev"},
				},
			},
		},
		{
			name:     "direct", // vulnerability in direct dependency
			version:  "1.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/basic-universe.yaml",
			requirements: []requirement{
				{
					name:    "bad",
					version: "^2.0.0",
				},
			},
		},
		{
			name:     "duplicates", // same package with vulns included multiple times
			version:  "1.1.1",
			system:   resolve.NPM,
			universe: "./fixtures/basic-universe.yaml",
			requirements: []requirement{
				{
					name:    "bad",
					version: "^1.0.0",
					typ:     aliasType("bad-aliased"),
				},
				{
					name:    "dependency",
					version: "^2.0.0",
					groups:  []string{"dev"},
				},
				{
					name:    "dependency",
					version: "^1.0.0",
					typ:     aliasType("dependency-v1"),
				},
			},
		},
		{
			name:     "different-pkgs", // same vuln in two different packages
			version:  "3.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/basic-universe.yaml",
			requirements: []requirement{
				{
					name:    "bad2",
					version: "^1.0.0",
				},
				{
					name:    "dependency",
					version: "^1.0.0",
				},
			},
		},
		{
			name:     "existing", // manifest package/version exists in universe already
			version:  "1.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/basic-universe.yaml",
			requirements: []requirement{
				{
					name:    "dependency",
					version: "^2.0.0",
					typ:     dep.NewType(dep.Opt),
				},
			},
		},
		{
			name:     "non-problem", // non-problem chains
			version:  "1.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/basic-universe.yaml",
			requirements: []requirement{
				{
					name:    "bad",
					version: "^1.0.0",
				},
				{
					name:    "dependency",
					version: "^3.0.0",
				},
			},
		},
		{
			name:     "diamond", // diamond dependency on vulnerable pkg
			version:  "1.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/diamond-universe.yaml",
			requirements: []requirement{
				{
					name:    "pkg",
					version: "^1.0.0",
				},
				{
					name:    "dep-one",
					version: "^1.0.0",
					groups:  []string{"dev"},
				},
			},
		},
		{
			name:     "complex", // more complex graph/vulnerability structure
			version:  "9.9.9",
			system:   resolve.NPM,
			universe: "./fixtures/complex-universe.yaml",
			requirements: []requirement{
				{
					name:    "alice",
					version: "^1.0.0",
					typ:     aliasType("chuck"),
				},
				{
					name:    "bob",
					version: "2.2.2",
				},
				{
					name:    "dave",
					version: "~3.3.3",
					groups:  []string{"dev"},
				},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			cl := clienttest.NewMockResolutionClient(t, tt.universe)
			var m manifest.Manifest
			m.Root = resolve.Version{
				VersionKey: resolve.VersionKey{
					PackageKey: resolve.PackageKey{
						Name:   tt.name,
						System: tt.system,
					},
					Version:     tt.version,
					VersionType: resolve.Concrete,
				},
			}
			m.Groups = make(map[manifest.RequirementKey][]string)
			m.Requirements = make([]resolve.RequirementVersion, len(tt.requirements))
			for i, req := range tt.requirements {
				m.Requirements[i] = resolve.RequirementVersion{
					VersionKey: resolve.VersionKey{
						PackageKey: resolve.PackageKey{
							Name:   req.name,
							System: tt.system,
						},
						Version:     req.version,
						VersionType: resolve.Requirement,
					},
					Type: req.typ,
				}
				m.Groups[manifest.MakeRequirementKey(m.Requirements[i])] = req.groups
			}

			res, err := resolution.Resolve(context.Background(), cl, m, tt.opts)
			if err != nil {
				t.Fatalf("error resolving: %v", err)
			}
			checkResult(t, res)
		})
	}
}
