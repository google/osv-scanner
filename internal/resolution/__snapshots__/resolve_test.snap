
[TestResolve/complex - 1]
complex 9.9.9
├─ reg|KnownAs="chuck"|Selector="" | alice@^1.0.0 1.0.1
│  └─ $1@^2.0.0
├─ 1: reg|Selector="" | bob@2.2.2 2.2.2
└─ reg|Selector="" | dave@~3.3.3 3.3.3
   ├─ $1@^2.2.2
   └─ reg|KnownAs="duck"|Selector="" | chuck@^2.0.0 2.0.0
      └─ $1@^2.0.1

---

[TestResolve/complex - 2]
[
  {
    "ID": "CMPLX-0000-0000",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "complex",
              "VersionType": 1,
              "Version": "9.9.9"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "alice",
              "VersionType": 1,
              "Version": "1.0.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  },
  {
    "ID": "CMPLX-1000-0000",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 2,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "complex",
              "VersionType": 1,
              "Version": "9.9.9"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "2.2.2",
                "Type": {}
              },
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 0,
                "To": 3,
                "Requirement": "~3.3.3",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "alice",
              "VersionType": 1,
              "Version": "1.0.1"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "bob",
              "VersionType": 1,
              "Version": "2.2.2"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "2.2.2",
                "Type": {}
              },
              {
                "From": 1,
                "To": 2,
                "Requirement": "^2.0.0",
                "Type": {}
              },
              {
                "From": 3,
                "To": 2,
                "Requirement": "^2.2.2",
                "Type": {}
              },
              {
                "From": 4,
                "To": 2,
                "Requirement": "^2.0.1",
                "Type": {}
              }
            ],
            "Children": null
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "dave",
              "VersionType": 1,
              "Version": "3.3.3"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "~3.3.3",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 3,
                "To": 2,
                "Requirement": "^2.2.2",
                "Type": {}
              },
              {
                "From": 3,
                "To": 4,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "4": {
            "Version": {
              "System": 3,
              "Name": "chuck",
              "VersionType": 1,
              "Version": "2.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 3,
                "To": 4,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 4,
                "To": 2,
                "Requirement": "^2.0.1",
                "Type": {}
              }
            ]
          }
        }
      }
    ]
  },
  {
    "ID": "CMPLX-2000-0000",
    "DevOnly": true,
    "Subgraphs": [
      {
        "Dependency": 4,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "complex",
              "VersionType": 1,
              "Version": "9.9.9"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "~3.3.3",
                "Type": {}
              }
            ]
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "dave",
              "VersionType": 1,
              "Version": "3.3.3"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "~3.3.3",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 3,
                "To": 4,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "4": {
            "Version": {
              "System": 3,
              "Name": "chuck",
              "VersionType": 1,
              "Version": "2.0.0"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 3,
                "To": 4,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  }
]
---

[TestResolve/diamond - 1]
diamond 1.0.0
├─ reg|Selector="" | pkg@^1.0.0 1.0.0
│  └─ 1: reg|Selector="" | bad@^1.0.0 1.1.1
└─ reg|Selector="" | dep-one@^1.0.0 1.0.0
   ├─ $1@^1.0.0
   └─ reg|Selector="" | dep-two@^1.0.0 1.0.0
      └─ $1@^1.0.0

---

[TestResolve/diamond - 2]
[
  {
    "ID": "DIA-000-000",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 3,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "diamond",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 0,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "pkg",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 1,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "dep-one",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 2,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 2,
                "To": 4,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 1,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 2,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 4,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          },
          "4": {
            "Version": {
              "System": 3,
              "Name": "dep-two",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 2,
                "To": 4,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 4,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          }
        }
      }
    ]
  }
]
---

[TestResolve/different-pkgs - 1]
different-pkgs 3.0.0
├─ reg|Selector="" | bad2@^1.0.0 1.0.0
└─ reg|Selector="" | dependency@^1.0.0 1.0.0
   └─ reg|Selector="" | bad@^1.0.0 1.1.1

---

[TestResolve/different-pkgs - 2]
[
  {
    "ID": "OSV-000-000",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 3,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "different-pkgs",
              "VersionType": 1,
              "Version": "3.0.0"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 2,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 2,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  },
  {
    "ID": "OSV-000-001",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "different-pkgs",
              "VersionType": 1,
              "Version": "3.0.0"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "bad2",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      },
      {
        "Dependency": 3,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "different-pkgs",
              "VersionType": 1,
              "Version": "3.0.0"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 2,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 2,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  }
]
---

[TestResolve/direct - 1]
direct 1.0.0
└─ reg|Selector="" | bad@^2.0.0 2.2.2

---

[TestResolve/direct - 2]
[
  {
    "ID": "OSV-000-001",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "direct",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "2.2.2"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  }
]
---

[TestResolve/duplicates - 1]
duplicates 1.1.1
├─ reg|KnownAs="bad-aliased"|Selector="" | bad@^1.0.0 1.1.1
├─ reg|Selector="" | dependency@^2.0.0 2.0.0
│  └─ reg|Selector="" | bad@^2.0.0 2.2.2
└─ reg|KnownAs="dependency-v1"|Selector="" | dependency@^1.0.0 1.0.0
   └─ reg|Selector="" | bad@^1.0.0 1.1.1

---

[TestResolve/duplicates - 2]
[
  {
    "ID": "OSV-000-000",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "duplicates",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      },
      {
        "Dependency": 5,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "duplicates",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 3,
                "To": 5,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "5": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 3,
                "To": 5,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  },
  {
    "ID": "OSV-000-001",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "duplicates",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      },
      {
        "Dependency": 4,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "duplicates",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "2.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 2,
                "To": 4,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "4": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "2.2.2"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 2,
                "To": 4,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      },
      {
        "Dependency": 5,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "duplicates",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "3": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 3,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 3,
                "To": 5,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "5": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 3,
                "To": 5,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  }
]
---

[TestResolve/existing - 1]
existing 1.0.0
└─ opt|Selector="" | dependency@^2.0.0 2.0.0
   └─ reg|Selector="" | bad@^2.0.0 2.2.2

---

[TestResolve/existing - 2]
[
  {
    "ID": "OSV-000-001",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 2,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "existing",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "2.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "2.2.2"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^2.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  }
]
---

[TestResolve/non-problem - 1]
non-problem 1.0.0
├─ 1: reg|Selector="" | bad@^1.0.0 1.1.1
└─ reg|Selector="" | dependency@^3.0.0 3.0.0
   └─ $1@*

---

[TestResolve/non-problem - 2]
[
  {
    "ID": "OSV-000-000",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "non-problem",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 0,
                "To": 2,
                "Requirement": "^3.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 2,
                "To": 1,
                "Requirement": "*",
                "Type": {}
              }
            ],
            "Children": null
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "3.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^3.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 2,
                "To": 1,
                "Requirement": "*",
                "Type": {}
              }
            ]
          }
        }
      }
    ]
  },
  {
    "ID": "OSV-000-001",
    "DevOnly": false,
    "Subgraphs": [
      {
        "Dependency": 1,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "non-problem",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 0,
                "To": 2,
                "Requirement": "^3.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              },
              {
                "From": 2,
                "To": 1,
                "Requirement": "*",
                "Type": {}
              }
            ],
            "Children": null
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "3.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 2,
                "Requirement": "^3.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 2,
                "To": 1,
                "Requirement": "*",
                "Type": {}
              }
            ]
          }
        }
      }
    ]
  }
]
---

[TestResolve/simple - 1]
simple 1.0.0
└─ reg|Selector="" | dependency@^1.0.0 1.0.0
   └─ reg|Selector="" | bad@^1.0.0 1.1.1

---

[TestResolve/simple - 2]
[
  {
    "ID": "OSV-000-000",
    "DevOnly": true,
    "Subgraphs": [
      {
        "Dependency": 2,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "simple",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  },
  {
    "ID": "OSV-000-001",
    "DevOnly": true,
    "Subgraphs": [
      {
        "Dependency": 2,
        "Nodes": {
          "0": {
            "Version": {
              "System": 3,
              "Name": "simple",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 2,
            "Parents": null,
            "Children": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "1": {
            "Version": {
              "System": 3,
              "Name": "dependency",
              "VersionType": 1,
              "Version": "1.0.0"
            },
            "Distance": 1,
            "Parents": [
              {
                "From": 0,
                "To": 1,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ]
          },
          "2": {
            "Version": {
              "System": 3,
              "Name": "bad",
              "VersionType": 1,
              "Version": "1.1.1"
            },
            "Distance": 0,
            "Parents": [
              {
                "From": 1,
                "To": 2,
                "Requirement": "^1.0.0",
                "Type": {}
              }
            ],
            "Children": null
          }
        }
      }
    ]
  }
]
---
